code/site/components/com_pages/dispatcher/router/interface.php

```diff
@@ -0,0 +1,104 @@
+<?php
+/**
+ * Joomlatools Pages
+ *
+ * @copyright   Copyright (C) 2018 Johan Janssens and Timble CVBA. (http://www.timble.net)
+ * @license     GNU GPLv3 <http://www.gnu.org/licenses/gpl.html>
+ * @link        https://github.com/joomlatools/joomlatools-pages for the canonical source repository
+ */
+
+/**
+ * Dispatcher Router Interface
+ *
+ * Provides route building and parsing functionality
+ *
+ * @author  Johan Janssens <https://github.com/johanjanssens>
+ * @package Koowa\Library\Dispatcher\Router
+ */
+interface ComPagesDispatcherRouterInterface
+{
+    /**
+     * Resolve the request
+     *
+     * @return false|KHttpUrl Returns the matched route or false if no match was found
+     */
+    public function resolve();
+
+    /**
+     * Generate a route
+     *
+     * @param string $path The path to generate a route for
+     * @param array @params Associative array of parameters to replace placeholders with.
+     * @return KHttpUrl Returns the generated route
+     */
+    public function generate($path, array $query = array());
+
+    /**
+     * Get the request path
+     *
+     * @return string
+     */
+    public function getPath();
+
+    /**
+     * Qualify a url
+     *
+     * Replace the url authority with the authority of the request url
+     *
+     * @param KHttpUrl $url The url to qualify
+     * @param bool $replace If the url is already qualified replace the authority
+     * @return KHttpUrl
+     */
+    public function qualifyUrl(KHttpUrl $url);
+
+    /**
+     * Set the response object
+     *
+     * @param KControllerResponseInterface $response A response object
+     * @return ComPagesDispatcherRouterInterface
+     */
+    public function setResponse(KControllerResponseInterface $response);
```

Comment:
@johanjanssens It doesn't feel quite right that the router package requires knowledge of the response object. This is very AD HOC as this gets used exclusively for the cannonical URL setting. Instead I would add a request only dependency here (setRequest/getRequest), which is what I would really expect and leave the cannonical URL setting outside the router package. I'd say that should happen here => https://github.com/joomlatools/joomlatools-pages/blob/master/code/site/components/com_pages/dispatcher/http.php#L59 at the time you first resolve the route.

Reply:
This is by design and intended. The role of a router and route resolvers is to parse the request and transform it in such a way that is understandable for the underlying application, by giving the router

The router receives the response object and through that also has access to the request, one of the router resolvers resolve the request.

If you follow your logic then:

1. All the code that now sits in the response object should be moved to the dispatcher. 
2. The dispatcher would always need to set a canonical, or if wouldn't need to make a decision to set it or not, which breaks encapsulation of the resolver and negates the resolver approach. 
3. If you only pass in the request a redirect resolver cannot be implemented easily either. Then the dispatcher would again be made more thick, need to make decisions if the returned route is a redirect or not, which it cannot easily do. Again you break encapsulation that way and you negate the resolver queue approach.